Allow Analytics to be removed as an available module without breaking Site Kit.

techanvil opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #5071 

## Relevant technical choices

- Removed the Tag Manager dependency on the Analytics JS bundle so the Tag Manager bundle still loads when Analytics is disabled.
- Updated `get_available_modules` to filter out modules which are missing any of the dependencies specified in the module's `depends_on` list. Practically speaking this means the Optimize module will also be removed if Analytics is disabled, as Optimize `depends_on` Analytics.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

aaemnnosttv commented on assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/Overview.test.js:
```diff
@@ -0,0 +1,120 @@
+import { getWidgetComponentProps } from '../../../../../googlesitekit/widgets/util';
+import Overview from './Overview';
+
+jest.mock( '../../../../../hooks/useBreakpoint', () => ( {
+	BREAKPOINT_SMALL: 'small',
+	useBreakpoint: () => 'tablet',
+} ) );
```
Would you please add some context as to why this is needed and why these specific values are provided?

techanvil replied:
Thanks for bringing this up. This has made me take stock, and take a different approach - see https://github.com/google/site-kit-wp/pull/5713/commits/62724fd634bdd46663667efc3dc6f29518eae6d6

aaemnnosttv commented on assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/Overview.test.js:
```diff
@@ -0,0 +1,120 @@
+	} );
+
+	it( 'should render the Search Funnel Overview, including the Activate Analytics CTA', async () => {
+		const { container, getByText } = await waitFor( () =>
```
If the `waitFor` is here due to registry state changes, you can wait for those using `waitForRegistry` returned by our `render` instead 👍

techanvil replied:
Thanks for pointing this out, I'll change to `waitForRegistry`.

Incidentally I had copied this `waitFor` usage from another test, so I have created a new issue to update the existing tests - https://github.com/google/site-kit-wp/issues/5734. Please take a look, feel free to close if you think it's unnecessary!

aaemnnosttv commented on assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/index.js:
```diff
@@ -76,7 +76,15 @@ const SearchFunnelWidget = ( {
 
 	const viewOnly = useViewOnly();
 
+	const isAnalyticsAvailable = useSelect( ( select ) =>
+		select( CORE_MODULES ).isModuleAvailable( 'analytics' )
+	);
+
 	const canViewSharedAnalytics = useSelect( ( select ) => {
+		if ( ! isAnalyticsAvailable ) {
+			return false;
+		}
+
```
This shouldn't be necessary because `canViewSharedModule` already checks for `module === null` internally. If anything, we should probably update that selector to use `isModuleAvailable` internally.

techanvil replied:
I think we still do need this check to avoid always returning `true` when in view-only mode...

aaemnnosttv commented on assets/js/modules/tagmanager/components/common/FormInstructions.js:
```diff
@@ -47,12 +47,19 @@ export default function FormInstructions( { isSetup } ) {
-	const analyticsPropertyID = useSelect( ( select ) =>
-		select( MODULES_ANALYTICS ).getPropertyID()
+	const analyticsModuleAvailable = useSelect( ( select ) =>
+		select( CORE_MODULES ).isModuleAvailable( 'analytics' )
```
I think this could be selected within the select for `analyticsPropertyID` below as the only other use already relies on `analyticsModuleActive` which should be sufficient.

techanvil replied:
The check further down for `analyticsModuleAvailable` is required in conjunction with `! analyticsModuleActive` - it's not sufficient to simply rely on `! analyticsModuleActive`.

This is because we only want to handle the "not active" case when the module _does_ exist - in this particular case, we only want to show the text containing "Activate the Google Analytics module" when the module exists to be activated, even while `analyticsModuleActive` returns `false`.

aaemnnosttv commented on tests/phpunit/integration/Core/Modules/ModulesTest.php:
```diff
@@ -62,6 +62,44 @@ function ( $instance ) {
+		$this->assertEqualSets(
+			array(
+				'adsense'            => 'Google\\Site_Kit\\Modules\\AdSense',
+				'analytics-4'        => 'Google\\Site_Kit\\Modules\\Analytics_4',
```
I would expect `analytics-4` to also not be present if `analytics` were not available as the two are not fully independent. I don't think anything for GA4 would be shown without Analytics available to cause a problem but we should probably define `analytics` as a dependency of `analytics-4`.

What should happen if `analytics-4` was filtered out but not `analytics`? Can they depend on each other? 😄

techanvil replied:
Did you mean by making use of `depends_on`, or by introducing a new approach for indicating module dependencies?

Making `analytics` a dependency of `analytics-4` via `depends_on` has the expected result of removing `analytics-4` when `analytics` is not available, so we could do that with minimal cost.

However, we can't make `analytics-4` a dependency of `analytics` via `depends_on`, as this results in the setup link for Analytics being disabled in normal usage, as `analytics-4` can't be setup independently:

![image](https://user-images.githubusercontent.com/18395600/186727985-33a5eb42-8c88-4dd2-b340-90c2de4fc7e8.png)

aaemnnosttv replied:
> Making `analytics` a dependency of `analytics-4` via `depends_on` has the expected result of removing `analytics-4` when `analytics` is not available, so we could do that with minimal cost.

`analytics-4` isn't usable without `analytics` so I think this makes sense.

If someone filtered out GA4 though I think other problems would arise that probably aren't accounted for here but we can't add the dependency to `analytics` as that would be a cyclical dependency as well. I'm less worried about this scenario though as that would be "doing it wrong" I think. If anything, maybe we could make it so that `analytics-4` is always preserved so long as `analytics` is in the list of available modules (essentially preventing this possibility). It's a rather extreme edge case IMO though so, not a big deal, just thinking out loud.

techanvil replied:
Cool, I've added `analytics` as a dependency of `analytics-4` via `depends_on`, so when filtering out `analytics`, `analytics-4` will be automatically removed too.

To be honest I don't think we should worry about the edge case of `analytics-4` being removed on its own. Analytics-4 is not exposed as an identifiable public facing module in its own right, and there's no actual reason to remove it in isolation - as you've said this would "doing it wrong", and on that basis I don't think we need to add logic for this scenario.

aaemnnosttv commented on tests/phpunit/integration/Core/Modules/ModulesTest.php:
```diff
@@ -62,6 +62,44 @@ function ( $instance ) {
+		add_filter(
+			'googlesitekit_available_modules',
+			function( $modules ) {
+				return array_filter(
+					$modules,
+					function( $module ) {
+						// Remove Analytics from the list of available modules.
+						return 'analytics' !== $module;
+					}
+				);
```
Let's provide a specific list here instead of using the filter otherwise this test will break when new modules are introduced.

techanvil replied:
I am of the opinion that we should keep this as-is. My view is this test is somewhat complimentary to `test_get_available_modules`. Please take a look at `test_get_available_modules` and you'll notice how the assertion is similar, and when we add new modules, we will need to update `test_get_available_modules` - at which point we'd update `test_get_available_modules__missing_dependency` too.

I think this would be a reasonable approach, but please take a look and see what you think.

aaemnnosttv commented on tests/phpunit/integration/Core/Modules/ModulesTest.php:
```diff
@@ -62,6 +62,44 @@ function ( $instance ) {
+
+		// Analytics is no longer present due to the filter above.
+		// Optimize is no longer present due to its dependency on Analytics.
+		$this->assertEqualSets(
```
This should use `assertSameSetsWithIndex` as it's an associative array. `assertEqualSets` treats them as lists and keys won't be considered in the comparison. That probably doesn't really make a difference here but the intention seems to be that it would, otherwise I would expect the assertion to use a list of values instead of keyed values.

techanvil replied:
Thanks for pointing that out. I had copied this from `test_get_available_modules`, so I have updated the usage there and in a number of other applicable places within this file.

techanvil replied:
Further on this - having updated to use `assertSameSetsWithIndex` I found the test suite failing on the [PHP (PHP 5.6, WordPress 4.7)](https://github.com/google/site-kit-wp/runs/8022066714?check_suite_focus=true#logs) workflow with an error:

> PHP Fatal error:  Call to undefined method Google\Site_Kit\Tests\Core\Modules\ModulesTest::assertSameSetsWithIndex() in /home/runner/work/site-kit-wp/site-kit-wp/tests/phpunit/integration/Core/Modules/ModulesTest.php on line 50

I have noticed that `assertSameSetsWithIndex` is not currently in use anywhere else in the codebase - I wonder if this is why?

I'll update to `assertEqualSetsWithIndex` which I can see in common use, but interested if you have any thoughts on this one.

aaemnnosttv replied:
Oops, that's probably what I meant. Looks like the `same` variant was only added in `5.6.0` where the `equals` ones have been around much longer. Thanks!

aaemnnosttv reviewed (changes requested):
Thanks @techanvil – this looks great overall, and thanks for all the added test coverage! I left a few comments along with a question or two.

Just to elaborate on one of my points, currently the `isModule*` selectors build on each other, so I would expect one to infer the others "below" it as there is no real reason why they should be independent.

- isModuleConnected: available + active + connected
- isModuleActive: available + active
- isModuleAvailable: available

Again, these should already work this way, but we shouldn't need to duplicate this in components.

techanvil commented on assets/js/components/setup/SetupUsingProxyWithSignIn.js:
```diff
@@ -384,9 +387,10 @@ export default function SetupUsingProxyWithSignIn() {
-												{ ! analyticsModuleActive && (
-													<ActivateAnalyticsNotice />
-												) }
+												{ analyticsModuleAvailable &&
```
Similarly, this check for `analyticsModuleAvailable` is required in conjunction with `! analyticsModuleActive` - it's not sufficient to simply rely on `! analyticsModuleActive`.

Again this is because we only want to handle the "not active" case when the module _does_ exist, in this case only showing the `ActivateAnalyticsNotice` when Analytics does exist to be activated

techanvil replied:
Similarly, this check for `analyticsModuleAvailable` is required in conjunction with `! analyticsModuleActive` - it's not sufficient to simply rely on `! analyticsModuleActive`.

Again this is because we only want to handle the "not active" case when the module _does_ exist, in this case only showing the `ActivateAnalyticsNotice` when Analytics does exist to be activated

techanvil replied:
Similarly, this check for `analyticsModuleAvailable` is required in conjunction with `! analyticsModuleActive` - it's not sufficient to simply rely on `! analyticsModuleActive`.

Again this is because we only want to handle the "not active" case when the module _does_ exist, in this case only showing the `ActivateAnalyticsNotice` when Analytics does exist to be activated

techanvil replied:
Similarly, this check for `analyticsModuleAvailable` is required in conjunction with `! analyticsModuleActive` - it's not sufficient to simply rely on `! analyticsModuleActive`.

Again this is because we only want to handle the "not active" case when the module _does_ exist, in this case only showing the `ActivateAnalyticsNotice` when Analytics does exist to be activated

techanvil commented on assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/Overview.js:
```diff
@@ -106,12 +114,18 @@ const Overview = ( {
-	const isNavigatingToReauthURL = useSelect( ( select ) =>
-		select( CORE_LOCATION ).isNavigatingTo( adminReauthURL )
-	);
+	const adminReauthURL = useSelect( ( select ) => {
```
Good shout, I've made the change :+1:

techanvil commented on assets/js/components/adminbar/AdminBarWidgets.js:
```diff
@@ -131,7 +134,9 @@ export default function AdminBarWidgets() {
 				) }
 
-				{ ( ! analyticsModuleConnected || ! analyticsModuleActive ) && (
+				{ ( ( analyticsModuleAvailable &&
```
Whoops - this comment has disappeared from the "Files changed" view and I've only just spotted it, hence addressing it last.

My answer to this is along the same lines as the comment below - this check for `analyticsModuleAvailable` is required in conjunction with `! analyticsModuleConnected || ! analyticsModuleActive`.

This is because we only want to handle the "not connected/active" case when the module _does_ exist, only showing the Analytics-related components when Analytics does exist to be connected/activated.

techanvil commented:
Hey @aaemnnosttv, I have merged `develop` to resolve conflicts and fixed subsequently broken tests.

While fixing up for the `AdminBarWidget` component I also spotted an out-of-date story which I have removed.

aaemnnosttv commented on assets/js/components/adminbar/AdminBarWidgets.js:
```diff
@@ -81,7 +84,9 @@ export default function AdminBarWidgets() {
 				) }
 
-				{ ( ! analyticsModuleConnected || ! analyticsModuleActive ) && (
+				{ ( ( analyticsModuleAvailable &&
+					! analyticsModuleConnected ) ||
+					! analyticsModuleActive ) && (
```
From the previous comments, I think the intention here is:
`available && ( ! connected || ! active )`

but this is 
`( available && ! connected ) || ! active )`

techanvil replied:
Good spot, thanks! I've fixed the condition :+1:

aaemnnosttv commented on assets/js/hooks/useBreakpoint.test.js:
```diff
@@ -50,18 +64,11 @@ describe( 'useBreakpoint', () => {
-						configurable: true,
-					}
-				);
+				setViewportWidth( viewportWidth );
```
This looks like a nice improvement – is it related to the other changes?

techanvil replied:
Yeah, I extracted `setViewportWidth` from here to reuse in `assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/Overview.test.js` elsewhere in this PR.

aaemnnosttv commented on assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/index.test.js:
```diff
@@ -0,0 +1,110 @@
+			{ registry }
+		);
+
+		await waitForElementToBeRemoved(
```
Is it not possible to use `waitForRegistry` here instead? If so, it would probably be worth leaving a comment to explain 👍

techanvil replied:
Thanks for the suggestion - `waitForRegistry` works here, no need to wait for a specific element to be removed. I've made the change here and elsewhere in the PR :+1:

aaemnnosttv commented on assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/index.test.js:
```diff
@@ -0,0 +1,110 @@
+			{ registry }
+		);
+
+		await waitForElementToBeRemoved(
```
Same here.

aaemnnosttv commented on includes/Modules/Tag_Manager.php:
```diff
@@ -505,12 +505,14 @@ protected function setup_assets() {
 						'googlesitekit-data',
 						'googlesitekit-datastore-site',
 						'googlesitekit-modules',
-						'googlesitekit-modules-analytics',
```
I think we should still probably add this conditionally if it is available. That way the code paths that run when it is available still run as today in terms of when the GA bundle is loaded. What do you think?

techanvil replied:
That sounds sensible. I have made this change, but in order to do so we need to allow the Tag Manager module to check whether the Analytics module is available - at present a `Module` has no real awareness of other `Modules`.

The approach I've taken is to pass in the list of available modules to the `get_assets` and `setup_assets` methods. However, it's arguably a bit brittle amending this shared method signature in this way just for the one instance of it. You can see the changes [here](https://github.com/google/site-kit-wp/pull/5713/files/1f73cf7b7bafc3acd6e80557c7660832348f0f79..baabbe5d7923279573bab3be802836afb45e9b3e).

I'm wondering if we should take a different approach:

1. Pass in the entire `Modules` instance to `get_assets` / `setup_assets` for more flexibility and call `$modules->module_exists( 'analytics' )` within `setup_assets` to check if Analytics is present.
2. Amend the `Module` class to hold a reference to the `Modules` instance, passing it into each `Module` when it's instantiated in `Modules::get_available_modules`. Call `$this->modules->module_exists( 'analytics' )` in `setup_assets`. This doesn't feel like a nice solution because it introduces a circular reference between `Modules` and each `Module`.
3. Use a filter instead, in a similar way to how the `googlesitekit_analytics_adsense_linked` filter is used to allow the presence of `AdSense` to affect the `Analytics` module.
4. Something else entirely.

Please take a look, interested to hear your thoughts on this.

aaemnnosttv replied:
Thanks – I agree, it would be better to find another way to avoid passing the list through to all those methods. After thinking about it a bit, I think using a filter would be simplest. Basically, the asset handle used for the dependency here would be filtered in by the Analytics module class. This way the handle would only be added if Analytics were available as only available modules are registered. I wouldn't make the whole list of dependencies filterable though, just the one for Analytics, which could default to `false` or anything else falsey such that we could filter it out to not register a non-existent asset as a dependency which would obviously break things.

One idea I was toying with here as a kind of asset-polymorphism would be to register a `no-op` script handle – this could be done using a `Script_Data` instance with no dependencies of its own, or any other args really. Essentially an empty handle. Then here we could do something like this:

```php
array(
	'googlesitekit-api',
	'googlesitekit-data',
	'googlesitekit-datastore-site',
	'googlesitekit-modules',
	'googlesitekit-vendor',
	apply_filters( 'googlesitekit-conditional-analytics-asset', 'googlesitekit-no-op' ),
)
```

That way, if Analytics were disabled, it would use the no-op asset which would technically be "enqueued" but otherwise do nothing. I'm not sure if this would really be better than using a filter without the no-op and just filtering out any falsey values from the list but just an idea.

techanvil replied:
Thanks @aaemnnosttv, I have reverted the changes to the signature for `get_assets` / `setup_assets` and used a filter instead. However I have taken a different approach to the filter than the one you proposed, I hope you agree it makes sense.

I noticed we already have this `googlesitekit_is_module_recoverable` filter in `Modules`:

https://github.com/google/site-kit-wp/blob/6b407087c078a557680f3a783cbc2ba5cbb04344/includes/Core/Modules/Modules.php#L391-L398

This is used as the basis of the `Module::is_recoverable` method:

https://github.com/google/site-kit-wp/blob/6b407087c078a557680f3a783cbc2ba5cbb04344/includes/Core/Modules/Module.php#L875-L884

I figure, especially as we already have this approach, providing a `googlesitekit_module_exists` filter would be generally useful and in keeping with the existing codebase. So I have added this to `Modules` and used the filter directly within `Tag_Manager::setup_assets`. Maybe we should add a `Module::module_exists( $slug )` method to encapsulate the filter usage, but I am not entirely sure it's needed.

Anyhow, the change is [here](https://github.com/google/site-kit-wp/pull/5713/files/701bbc18de22e63308fd11efba4d0fb1dfa31aa6..6b407087c078a557680f3a783cbc2ba5cbb04344), please see what you think.

aaemnnosttv replied:
Thanks, this sounds like a good solution to me 👍 Nice work!

aaemnnosttv commented on tests/js/viewport-width-utils.js:
```diff
@@ -0,0 +1,28 @@
+/**
```
Please add a file header 🙂

techanvil replied:
Ah whoops! Done, thanks :+1:

aaemnnosttv reviewed (commented):
Thanks for your patience here @techanvil ! This looks quite close, just a handful of details left to address 👍

aaemnnosttv reviewed (commented):
Thanks @techanvil – this is almost there. I think the only thing left is to simplify the changes around conditionally enqueuing the Analytics asset. See my reply below.

aaemnnosttv commented on tests/phpunit/integration/Modules/Tag_ManagerTest.php:
```diff
@@ -360,6 +361,76 @@ public function test_get_datapoints() {
+			$dependency->deps
+		);
+
+		wp_deregister_script( 'googlesitekit-modules-tagmanager' );
```
Is this necessary? I would expect this to not be persisted between tests . If it does need to be cleaned up, let's do that in a before/after method to ensure it's cleaned up properly rather than in the test where it won't run if the test fails which could cause other tests to fail as a result.

techanvil replied:
It is necessary, unfortunately. Good shout though, I've moved the `wp_deregister_script` calls to to a `tear_down` method.

aaemnnosttv commented on tests/phpunit/integration/Modules/Tag_ManagerTest.php:
```diff
@@ -360,6 +361,76 @@ public function test_get_datapoints() {
+
+		$this->assertEquals( $context->url( 'dist/assets/' ) . 'js/googlesitekit-modules-tagmanager.js', $dependency->src );
+
+		// Verify the 'googlesitekit-module-analytics' asset is not a dependency.
```
How about we use an assertion for this instead?

techanvil replied:
Hmm, if I am following you correctly do you mean something like this?

```php
		$this->assertNotContains(
			'googlesitekit-module-analytics',
			$dependency->deps
		);
```

I would prefer to keep the current assertion as it's more robust, guarding against `$dependency->deps` containing an unexpected value (or being empty etc) via this execution path.

However I have gone ahead and added the `assertNotContains` as well, so as to state the intent in code rather than a comment.

aaemnnosttv commented on includes/Core/Modules/Modules.php:
```diff
@@ -379,6 +379,15 @@ function ( $data ) {
 			}
 		);
 
+		add_filter(
+			'googlesitekit_module_exists',
+			function ( $exists, $slug ) {
+				return $this->module_exists( $slug );
+			},
+			10,
+			2
+		);
```
Please add coverage for this in `ModulesTest::test_register`.

techanvil replied:
Good shout, I've added some test coverage :+1:

aaemnnosttv reviewed (commented):
Thanks @techanvil – I think this should be the last round, and there's not much left to address. Will try to get this merged asap once it's ready again. Nice work!

aaemnnosttv reviewed (approved):
Excellent work, thanks @techanvil !
